### PR TITLE
Calculate text positions for constant strings

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description
Constant strings for text annotations of shapes layers can be specified with something like `shapes.text = {'string': {'constant': 'hello'}}`. However, when we do that I get the following error.

<details>

```
WARNING: Traceback (most recent call last):
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/app/backends/_qt.py", line 903, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/util/event.py", line 453, in __call__
    self._invoke_callback(cb, event)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/util/event.py", line 471, in _invoke_callback
    _handle_exception(self.ignore_callback_errors,
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/util/event.py", line 469, in _invoke_callback
    cb(event)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/scene/canvas.py", line 218, in on_draw
    self._draw_scene()
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/scene/canvas.py", line 277, in _draw_scene
    self.draw_visual(self.scene)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/scene/canvas.py", line 315, in draw_visual
    node.draw()
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/scene/visuals.py", line 103, in draw
    self._visual_superclass.draw(self)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/visuals/visual.py", line 605, in draw
    v.draw()
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/scene/visuals.py", line 103, in draw
    self._visual_superclass.draw(self)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/visuals/visual.py", line 442, in draw
    if self._prepare_draw(view=self) is False:
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/visuals/text/text.py", line 570, in _prepare_draw
    assert pos.shape[0] == self._vertices.size == len(_rot)
AssertionError
```

</details>

That's because we [currently skip the calculation of text positions when a shapes layer's text string is constant](https://github.com/napari/napari/blob/5ca30ea2a4ef73fb2a2a1da6a93430ca286d1e88/napari/layers/shapes/shapes.py#L1551). We have a [similar but more specific check to handle that optimization on the vispy side of things](https://github.com/napari/napari/blob/5ca30ea2a4ef73fb2a2a1da6a93430ca286d1e88/napari/_vispy/utils/text.py#L69), so I think we can simply remove that code in the shapes layer to fix the above bug.

## Type of change
- [x] Bug-fix (non-breaking change which fixes an issue)

# References
Related to issue #5378, since specifying `string: {'constant': '200'}` should be the official workaround for that issue.

# How has this been tested?
- [ ] modified tests with coordinate checks pass with my change
- [ ] all existing tests pass with my change

### CODE REVIEW

**Comment by andy-sweet** [comment]
@maweigert : since you added the short circuit in https://github.com/napari/napari/pull/3867, I wanted to check that removing makes some sense - see the PR description for the reasoning.

File: napari/_vispy/_tests/test_vispy_shapes_layer.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 def test_remove_selected_with_derived_text():
     """See https://github.com/napari/napari/issues/3504"""
+    np.random.seed(0)
```
Comment (andy-sweet) [inline napari/_vispy/_tests/test_vispy_shapes_layer.py:9]: As I was modifying these tests and know that random shapes can cause flaky tests, I seeded these to make them reproducible.

**Comment by maweigert** [comment]
> @maweigert : since you added the short circuit in #3867, I wanted to check that removing makes some sense - see the PR description for the reasoning.

Sure! I am not entirely understanding why the short circuit would have been active for constant strings, as it should only apply for empty strings (in which case it  shaves of a lot of layer creation time). Maybe there is a way to keep it for truely empty strings?

**Comment by andy-sweet** [comment]
> Sure! I am not entirely understanding why the short circuit would have been active for constant strings, as it should only apply for empty strings (in which case it shaves of a lot of layer creation time). Maybe there is a way to keep it for truely empty strings?

The main logic is as follows.

1. `compute_text_coords` is only called by `Shapes._view_text_coords`.
2. `Shapes._view_text_coords` is only called by `update_text` (in `napari._vispy.utils.text`)
3. `update_text` already handles this optimization (in a more correct way)

So I don't think removing it will cause any performance degradation to napari itself.

There are a couple of complications/details to this though.

`TextManager.values` is only intended to be broadcastable to an 1D array of strings (that should be better documented!). In the case that the string is a constant we decided to only store that scalar constant for efficiency (instead of an array with the length of the number of shapes). That's why your check isn't correct in this context - `self.text.values.shape == ()` evaluates to true if `values` is a scalar.

If I remember correctly, vispy does not handle empty string/coords for text visuals. So we workaround that altogether in `update_text`. I should probably write a bug report in vispy and maybe submit a fix instead of working around it in napari.

Sorry, this is more complicated than it should be! I could look at correcting the check, but my initial response was to remove it.

**Comment by maweigert** [comment]
Sounds good! I was just trying to remember why I put the check in in the first place :)

**Comment by andy-sweet** [comment]
@kevinyamauchi : assigned you as a reviewer due to some of your previous work here. This is fairly small and there's not too much rush on this.

**Comment by kevinyamauchi** [comment]
Thanks for the fix, @andy-sweet ! I can have a look today.

**Review by kevinyamauchi — APPROVED** [review]
This looks good to me. Thanks for the fix and tests @andy-sweet !

**Comment by andy-sweet** [comment]
Thanks, I'll merge this after 24 hours unless there are any objections.

**Pull request merged**